### PR TITLE
Fix cumsum error for tensors with zero elements

### LINKED ISSUE #31515: cumsum backward for zero-length zensors
## 🐛 Bug

Calling `cumsum` on a zero-dimension tensor raises an error in backward.
The error occurs for both `cumsum` and `cumprod` and on both CPU and GPU

## To Reproduce

```python
import torch

x = torch.zeros(2, 0, requires_grad=True)
xx = x.cumsum(dim=-1).sum()
xx.backward()
```

Output:
```
Traceback (most recent call last):
  File "narrow_test.py", line 6, in <module>
    xx.backward()
  File "xxx/torch/tensor.py", line 166, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph)
  File "xxx/torch/autograd/__init__.py", line 99, in backward
    allow_unreachable=True)  # allow_unreachable flag
RuntimeError: start (0) + length (1) exceeds dimension size (0).
```

## Expected behavior

I'm not sure if the gradient calculation makes any sense in this edge case, but at least the error message should be more clear.

## Environment

```
Collecting environment information...
PyTorch version: 1.3.1
Is debug build: No
CUDA used to build PyTorch: 10.1.243

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.2.2

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: TITAN RTX
GPU 1: TITAN RTX
GPU 2: TITAN RTX
GPU 3: TITAN RTX
GPU 4: TITAN RTX
GPU 5: TITAN RTX

Nvidia driver version: 430.50
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.17.2
[pip] torch==1.3.1
[pip] torchvision==0.4.2
[conda] blas                      1.0                         mkl
[conda] mkl                       2019.4                      243
[conda] mkl-service               2.3.0            py37he904b0f_0
[conda] mkl_fft                   1.0.15           py37ha843d7b_0
[conda] mkl_random                1.1.0            py37hd6b4f25_0
[conda] pytorch                   1.3.1           py3.7_cuda10.1.243_cudnn7.6.3_0    pytorch
[conda] torchvision               0.4.2                py37_cu101    pytorch
```

cc @ezyang @SsnL @albanD @zou3519 @gqchen

### PR DESCRIPTION
**Pull request by leedtan**
Currently `cumsum` crashes for tensors with non-empty dimensions but with zero elements, which could happen when some dimension is zero. This commit fixes the error by checking both `dim()` and `numel()` in cumsum backward

Fixes #31515

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -10416,6 +10416,11 @@ def test_cumsum(self, device):
                                              [0, 0, 0],
                                              [1, 2, 3]]))
 
+        # Check that cummulative sum over a zero length dimension doesn't crash on backprop
+        zeroNumelTensor = torch.zeros(2, 0, requires_grad=True)
+        integrated = zeroNumelTensor.cumsum(dim=-1).sum()
+        integrated.backward()
+
```
Comment (mrshenli) [inline test/test_torch.py]: Let's verify the dimension of the grad (`integrated .grad`) should match.

**Review by mrshenli — CHANGES_REQUESTED** [review]
Thanks for contributing! Some general comments:

1. The title of the PR needs to be more descriptive. This title will be the git commit title. Think from a reader's perspective, what title will make it easier for a reader to understand what this PR is doing.
2. Same applies to the PR description, it will be the commit message.
3. You can add "closes #31515" to the PR description, it will then automatically close the issue when this PR is merged.
4. Make sure you updated the internal diff description and title accordingly as well.

File: test/test_torch.py [hunk 2]
```diff
@@ -10416,6 +10416,11 @@ def test_cumsum(self, device):
                                              [0, 0, 0],
                                              [1, 2, 3]]))
 
+        # Check that cummulative sum over a zero length dimension doesn't crash on backprop
+        zeroNumelTensor = torch.zeros(2, 0, requires_grad=True)
```
[the remark this answers is not in the document]
Comment (leedtan) [reply test/test_torch.py]: I copied nearby lines, such as:
        aRes = torch.cumsum(a, 1)
        bRes = torch.cumsum(b, 1)

But if thats the chosen practices, I also see examples a few hundred lines above that are snake_case so I'll make the change 
"
                empty_strided = torch.empty_strided(shape, strides, device=device)
                as_strided = torch.empty(empty_strided.storage().size(),
                                         device=device).as_strided(shape, strides)
"
Should I also replace aRes to a_res?

File: test/test_torch.py [hunk 3]
```diff
@@ -10416,6 +10416,11 @@ def test_cumsum(self, device):
...
+        integrated = zeroNumelTensor.cumsum(dim=-1).sum()
```
[the remark this answers is not in the document]
Comment (leedtan) [reply test/test_torch.py]: What do you mean cumsum on a scalar? I'm going to interpret this as "put up more tests" and I will add tests like shape: (2, 1, 4) and shape: (0, 2, 3) and try out cumsum on (0, 1, 2) on all of the above shapes.

**Review by mrshenli — APPROVED** [review]
LGTM! Thanks for fixing!

**Review by mrshenli — CHANGES_REQUESTED** [review]
Lint failures are real, please address.

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit fce6a06b:


**None of the build failures appear to be your fault**.

* **1/1** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=b522a8e1ff8a531c4ac75a3551b99d5b40125cf0&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base b522a8e1 since Dec 27
  <details>
  <summary>
  You may want to rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  Since your merge base is older than `viable/strict`, run these commands:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  


---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

**Pull request closed without merging**